uqcsbot/scripts/leet.py

```diff
@@ -0,0 +1,143 @@
+from uqcsbot import bot, Command
+from http import HTTPStatus
+from uqcsbot.utils.command_utils import loading_status
+import json
+import requests
+import random
+from slackblocks import Attachment, SectionBlock
+
+LC_DIFFICULTY_MAP = ["easy", "medium", "hard"]  # leetcode difficulty is 1,2,3, need to map
+HR_DS_API_LINK = ("https://www.hackerrank.com/rest/contests/master/tracks/" +
+                  "data-structures/challenges?limit=200")
+HR_ALG_API_LINK = ("https://www.hackerrank.com/rest/contests/master/tracks/" +
+                   "algorithms/challenges?limit=200")
+LC_API_LINK = 'https://leetcode.com/api/problems/all/'
+
+
+COLORS = {"easy": "#5db85b",
+          "medium": "#f1ad4e",
+          "hard": "#d9534f"}
+
+
+@bot.on_command('leet')
+@loading_status
+def handle_leet(command: Command) -> None:
+    """
+    `!leet [`easy` | `medium` | `hard`] - Retrieves a set of questions from online coding
+    websites, and posts in channel with a random question from this set. If a difficulty
+    is provided as an argument, the random question will be restricted to this level of
+    challenge. Else, a random difficulty is generated to choose.
+    """
+    was_random = True  # Used for output later
+
+    if command.has_arg():
+        if (command.arg not in {"easy", "medium", "hard"}):
+            bot.post_message(command.channel_id, "Usage: !leet [`easy` | `medium` | `hard`]")
+            return
+        else:
+            difficulty = command.arg.lower()
+            was_random = False
+    else:
+        difficulty = random.choice(LC_DIFFICULTY_MAP)  # No difficulty specified, randomly generate
+
+    # List to store questions collected
+    questions = []
+
+    # Go fetch questions from APIs
+    collect_questions(questions, difficulty)
+    selected_question = select_question(questions)  # Get a random question
+
+    # If we didn't find any questions for this difficulty, try again, probably timeout on all 3
+    if (len(selected_question) == 0):
+        bot.post_message(command.channel_id,
+                         "Hmm, the internet pipes are blocked. Try that one again.")
+        return
+
+    # Leetcode difficulty colors
+    color = COLORS[difficulty]
+
+    if (was_random):
+        title_text = f"Random {difficulty} question generated!"
+    else:
+        # Style this a bit nicer
+        difficulty = difficulty.title()
+        title_text = f"{difficulty} question generated!"
+
+    difficulty = difficulty.title()  # If we haven't already (i.e. random question)
+
+    msg_text = f"Here's a new question for you! <{selected_question[1]}|{selected_question[0]}>"
+
+    bot.post_message(command.channel_id, text=title_text,
+                     attachments=[Attachment(SectionBlock(msg_text), color=color)._resolve()])
+
+
+def select_question(questions: list):
+    """
+    Small helper method that selects a question from a list randomly
+    """
+    if (len(questions) == 0):
+        return ()
```

Comment:
Prefer returning `None` over an empty tuple (this means the check on return val should be `is None` rather than `len() == 0`)

Reply:
fixed